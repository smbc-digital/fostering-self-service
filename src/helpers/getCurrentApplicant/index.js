import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant