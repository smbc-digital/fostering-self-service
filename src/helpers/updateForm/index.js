import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}