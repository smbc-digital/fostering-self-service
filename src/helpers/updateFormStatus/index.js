export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}