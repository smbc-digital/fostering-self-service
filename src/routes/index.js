import { StageName } from 'helpers'
import Start from 'components/Pages/Start'
import { KnownByAnotherName, MoreAboutYou } from 'components/Pages/HomeVisit/1-TellUsAboutYou'
import { AreYouEmployed, EmploymentDetails } from 'components/Pages/HomeVisit/2-YourEmploymentDetails'
import LanguagesSpokenInYourHome from 'components/Pages/HomeVisit/3-LanguagesSpokenInYourHome'
import { AreYouMarried, MarriageDate, MovedInTogetherDate } from 'components/Pages/HomeVisit/4-PartnershipStatus'
import { HaveYouPreviouslyApplied } from 'components/Pages/HomeVisit/5-YourFosteringHistory'
import YourHealth from 'components/Pages/HomeVisit/6-YourHealth'
import { AboutYourInterest, TypesOfFostering } from 'components/Pages/HomeVisit/7-InterestInFostering'
import { AnyUnderSixteens, AboutUnderSixteens, AnyOverSixteens, AboutOverSixteens } from 'components/Pages/HomeVisit/8-ChildrenAwayFromHome'
import { AnyPeopleInYourHousehold, PeopleInYourHousehold, DoYouHaveAnyPets } from 'components/Pages/HomeVisit/10-YourHousehold'
import { YourGpDetails } from 'components/Pages/Application'
import { FamilyReference, FirstPersonalReference, SecondPersonalReference } from 'components/Pages/Application/3-YourReferences'
import { DoYouKnowCouncillors } from 'components/Pages/Application/4-CouncillorsOrEmployees'

export const START_PAGE = '/fostering/your-fostering-journey'
export const KNOWN_BY_ANOTHER_NAME = '/fostering/known-by-another-name'
export const MORE_ABOUT_YOU = '/fostering/more-about-you'
export const ARE_YOU_EMPLOYED = '/fostering/are-you-employed'
export const EMPLOYMENT_DETAILS = '/fostering/employment-details'
export const LANGUAGES_SPOKEN_IN_YOUR_HOME = '/fostering/languages-spoken-in-your-home'
export const PARTNERSHIP_STATUS = '/fostering/your-partnership-status'
export const DATE_OF_MARRIAGE_OR_PARTNERSHIP = '/fostering/date-of-marriage-or-partnership'
export const DATE_OF_MOVE_IN_TOGETHER = '/fostering/date-of-move-in'
export const FOSTERING_HISTORY = '/fostering/your-fostering-history'
export const ABOUT_YOUR_HEALTH = '/fostering/about-your-health'
export const YOUR_INTEREST_IN_FOSTERING = '/fostering/your-interest-in-fostering'
export const TYPES_OF_FOSTERING = '/fostering/types-of-fostering'
export const YOUR_HOUSEHOLD = '/fostering/your-household'
export const WHO_LIVES_IN_YOUR_HOME = '/fostering/who-lives-in-your-home'
export const DO_YOU_HAVE_ANY_PETS = '/fostering/do-you-have-any-pets'
export const CHILDREN_UNDER_SIXTEEN_LIVING_AWAY = '/fostering/children-under-sixteen-living-away'
export const ABOUT_CHILDREN_UNDER_SIXTEEN_LIVING_AWAY = '/fostering/about-children-under-sixteen-living-away'
export const CHILDREN_OVER_SIXTEEN_LIVING_AWAY = '/fostering/children-over-sixteen-living-away'
export const ABOUT_CHILDREN_OVER_SIXTEEN_LIVING_AWAY = '/fostering/about-children-over-sixteen-living-away'
export const ABOUT_YOUR_GP = '/fostering/about-your-gp'
export const FAMILY_REFERENCE = '/fostering/family-reference'
export const FIRST_PERSONAL_REFERENCE = '/fostering/first-personal-reference'
export const SECOND_PERSONAL_REFERENCE = '/fostering/second-personal-reference'
export const RELATIONSHIP_TO_COUNCIL_EMPLOYEES = '/fostering/relationship-to-council-employees'
export const ABOUT_RELATIONSHIP_TO_COUNCIL_EMPLOYEES = '/fostering/about-relationship-to-council-employees'

export const RouterConfig = [
    {
        path: START_PAGE,
        component: Start
    },
    {
        path: KNOWN_BY_ANOTHER_NAME,
        component: KnownByAnotherName,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: MORE_ABOUT_YOU,
        component: MoreAboutYou,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: ARE_YOU_EMPLOYED,
        component: AreYouEmployed,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: EMPLOYMENT_DETAILS,
        component: EmploymentDetails,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: LANGUAGES_SPOKEN_IN_YOUR_HOME,
        component: LanguagesSpokenInYourHome,
        stage: StageName.HomeVisit
    },
    {
        path: PARTNERSHIP_STATUS,
        component: AreYouMarried,
        stage: StageName.HomeVisit,
        jointApplicantOnly: true
    },
    {
        path: DATE_OF_MARRIAGE_OR_PARTNERSHIP,
        component: MarriageDate,
        stage: StageName.HomeVisit,
        jointApplicantOnly: true
    },
    {
        path: DATE_OF_MOVE_IN_TOGETHER,
        component: MovedInTogetherDate,
        stage: StageName.HomeVisit,
        jointApplicantOnly: true
    },
    {
        path: FOSTERING_HISTORY,
        component: HaveYouPreviouslyApplied,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: ABOUT_YOUR_HEALTH,
        component: YourHealth,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: YOUR_INTEREST_IN_FOSTERING,
        component: AboutYourInterest,
        stage: StageName.HomeVisit
    },
    {
        path: TYPES_OF_FOSTERING,
        component: TypesOfFostering,
        stage: StageName.HomeVisit
    },
    {
        path: YOUR_HOUSEHOLD,
        component: AnyPeopleInYourHousehold,
        stage: StageName.HomeVisit
    },
    {
        path: WHO_LIVES_IN_YOUR_HOME,
        component: PeopleInYourHousehold,
        stage: StageName.HomeVisit
    },
    {
        path: DO_YOU_HAVE_ANY_PETS,
        component: DoYouHaveAnyPets,
        stage: StageName.HomeVisit
    },
    {
        path: CHILDREN_UNDER_SIXTEEN_LIVING_AWAY,
        component: AnyUnderSixteens,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: ABOUT_CHILDREN_UNDER_SIXTEEN_LIVING_AWAY,
        component: AboutUnderSixteens,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: CHILDREN_OVER_SIXTEEN_LIVING_AWAY,
        component: AnyOverSixteens,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: ABOUT_CHILDREN_OVER_SIXTEEN_LIVING_AWAY,
        component: AboutOverSixteens,
        stage: StageName.HomeVisit,
        bothApplicants: true
    },
    {
        path: ABOUT_YOUR_GP,
        component: YourGpDetails,
        stage: StageName.Application,
        bothApplicants: true
    },
    {
        path: FAMILY_REFERENCE,
        component: FamilyReference,
        stage: StageName.Application
    },
    {
        path: FIRST_PERSONAL_REFERENCE,
        component: FirstPersonalReference,
        stage: StageName.Application
    },
    {
        path: SECOND_PERSONAL_REFERENCE,
        component: SecondPersonalReference,
        stage: StageName.Application
    },
    {
        path: RELATIONSHIP_TO_COUNCIL_EMPLOYEES,
        component: DoYouKnowCouncillors,
        stage: StageName.Application,
        bothApplicants: true
    }
]