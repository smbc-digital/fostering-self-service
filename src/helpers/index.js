import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}