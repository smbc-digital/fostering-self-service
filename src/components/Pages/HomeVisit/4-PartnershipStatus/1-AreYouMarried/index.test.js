import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import AreYouMarried from './index'
import { DATE_OF_MARRIAGE_OR_PARTNERSHIP, DATE_OF_MOVE_IN_TOGETHER, FOSTERING_HISTORY } from 'routes'

describe('AreYouMarried', () => {

    beforeEach(() => {
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            marriedOrInACivilPartnership: {
                value: undefined,
                isValid: false
            },
            statuses: {
                yourPartnershipStatus: 0
            },
            onChangeStatus: jest.fn()
        })
    }) 

    it('should push to next page, when married', () => {
        // Arrange
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            marriedOrInACivilPartnership: {
                value: 'true',
                isValid: true
            },
            statuses: {
                yourPartnershipStatus: 0
            },
            onChangeStatus: jest.fn()
        })
        const history = {
            push: jest.fn()
        }
        const wrapper = mount(<AreYouMarried history={history} />)

        // Act
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(DATE_OF_MARRIAGE_OR_PARTNERSHIP)
    })

    it('should push to next page, when not married', () => {
        // Arrange
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            marriedOrInACivilPartnership: {
                value: 'false',
                isValid: true
            },
            statuses: {
                yourPartnershipStatus: 0
            },
            onChangeStatus: jest.fn()
        })
        const history = {
            push: jest.fn()
        }
        const wrapper = mount(<AreYouMarried history={history} />)

        // Act
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(DATE_OF_MOVE_IN_TOGETHER)
    })

    it('should push to next page', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const wrapper = mount(<AreYouMarried history={history} />)

        // Act
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(FOSTERING_HISTORY)
    })

    describe('snapshot', () => {

        it('should render correctly', () => {
            // Act
            const tree = renderer
                .create(<AreYouMarried history={{}} />)
                .toJSON()

            // Assert
            expect(tree).toMatchSnapshot()
        })
    })
})