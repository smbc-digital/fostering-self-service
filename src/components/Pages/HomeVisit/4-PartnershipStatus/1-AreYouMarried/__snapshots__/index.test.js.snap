// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreYouMarried snapshot should render correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your partnership status
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="no-margin-bottom"
    >
      Are you married or in a civil partnership?
    </h3>
    <fieldset>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="are-you-married-yes"
          name="marriedOrInACivilPartnership"
          onChange={[Function]}
          type="radio"
          value="true"
        />
        <label
          htmlFor="are-you-married-yes"
        >
          Yes
        </label>
      </div>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="are-you-married-no"
          name="marriedOrInACivilPartnership"
          onChange={[Function]}
          type="radio"
          value="false"
        />
        <label
          htmlFor="are-you-married-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
  </form>,
]
`;
